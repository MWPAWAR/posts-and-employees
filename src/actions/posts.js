import fetchPosts from '../apis/posts';
import { FETCH_POSTS, FETCH_POSTS_SUCCESS, FETCH_POSTS_FAILURE } from '../constants/action-types';

export default () => dispatch => {
  dispatch({ type: FETCH_POSTS });

  return fetchPosts()
    .then(response => {
      dispatch({ type: FETCH_POSTS_SUCCESS, payload: { data: response.data } });
    })
    .catch(() => {
      dispatch({ type: FETCH_POSTS_FAILURE });
    })
};
