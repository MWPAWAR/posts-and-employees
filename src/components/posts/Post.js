import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';

import Card from '../core/card/Card';
import Button, { BUTTON_TYPES } from '../core/button/Button';

const Post = ({
  id,
  body,
  title
}) => (
  <Card
    title={title}
    body={body}
  >
    <Link to={`/posts/${id}`}>
      <Button
        type={BUTTON_TYPES.LINK}
        text={<FormattedMessage id="app.posts.btn-view-label" />}
      />
    </Link>
  </Card>
);

Post.propTypes = {
  id: PropTypes.number.isRequired,
  body: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired
};

export default Post;
