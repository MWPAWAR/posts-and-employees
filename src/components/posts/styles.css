.root {
  width: 20rem;
  margin: 1rem;
}

.heading {
  margin: 1rem 0;
  text-align: center;
}

.container {
  display: flex;
  flex-wrap: wrap;
}

.post-details {
  margin: 1rem;
  width: 100%;
}
