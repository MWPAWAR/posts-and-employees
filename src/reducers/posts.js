import { FETCH_POSTS, FETCH_POSTS_SUCCESS, FETCH_POSTS_FAILURE } from '../constants/action-types';

const INITIAL_STATE = { isError: false };

export default (state = INITIAL_STATE, action) => {
  const { type, payload } = action;

  switch (type) {
    case FETCH_POSTS:
      return { ...state };
    case FETCH_POSTS_SUCCESS:
      return { ...state, ...payload, isError: false };
    case FETCH_POSTS_FAILURE:
      return { ...state, isError: true };
    default:
      return state;
  }
};
