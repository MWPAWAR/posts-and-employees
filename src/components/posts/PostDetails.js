import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { FormattedMessage } from 'react-intl';

import Card from '../core/card/Card';
import DataUnavailable from '../core/error/Error';
import DetailsRow from '../core/details-row/DetailsRow';
import getPostById from '../../selectors/post';
import './styles.css';

const PostDetails = ({ post }) => (
  post
    ? (
      <Card className="post-details" title={post.title}>
        <DetailsRow
          label={<FormattedMessage id="app.posts.post.description" />}
          value={post.body}
        />
      </Card>
    ) : (
      <DataUnavailable
        message={<FormattedMessage id="app.error-no-data" />}
      />
    )
);

const mapStateToPros = (state, props) => ({
  post: getPostById(state, props.match.params.postId)
});

PostDetails.propTypes = {
  post: PropTypes.shape({
    title: PropTypes.string.isRequired,
    body: PropTypes.string.isRequired,
  })
}

export { PostDetails }; // exporting unconnected component from here for teting.

export default connect(mapStateToPros)(PostDetails);
