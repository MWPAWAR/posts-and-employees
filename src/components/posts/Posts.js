import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { FormattedMessage } from 'react-intl';

import fetchData from '../../actions/posts';
import getPosts from '../../selectors/posts';
import Post from './Post';
import withSpinner from '../hocs/with-spinner';
import styles from './styles.module.css';

const Posts = ({ posts }) => (
  <>
    <h2 className={styles.heading}>
      <FormattedMessage id="app.posts.label" />
    </h2>
    <div className={styles.container}>
      {posts.map((post) => (
        <Post id={post.id} title={post.title} body={post.body} />
      ))}
    </div>
  </>
);

Posts.propTypes = {
  posts: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      body: PropTypes.string.isRequired
    })
  )
}

const mapStateToProps = (state) => {
  const posts = getPosts(state);
  const isDataAvailable = posts.length > 0;

  return {
    posts,
    isDataAvailable,
  };
};

export default connect(mapStateToProps, { fetchData })(withSpinner(Posts));
