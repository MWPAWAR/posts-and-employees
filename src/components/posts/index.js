import Post from './Post';
import PostDetails from './PostDetails';
import Posts from './Posts';

export { Post, PostDetails, Posts };
