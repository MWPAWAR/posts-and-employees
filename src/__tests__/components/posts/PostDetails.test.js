import React from 'react';
import { shallow } from 'enzyme';

import { PostDetails } from '../../../components/posts/PostDetails';
import postMock from '../../mocks/post.json';

describe('PostDetails', () => {
  test('renders data unavailable message if post is undefined', () => {
    const component = shallow(<PostDetails />);
    expect(component).toMatchSnapshot();
  });

  test('renders correctly with default data', () => {
    const component = shallow(<PostDetails post={postMock} />);
    expect(component).toMatchSnapshot();
  });
});
