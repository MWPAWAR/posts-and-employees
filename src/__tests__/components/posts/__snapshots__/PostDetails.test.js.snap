// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDetails renders correctly with default data 1`] = `
<Card
  className="postDetails"
  title="Welcome to Posts"
>
  <DetailsRow
    label={
      <FormattedMessage
        id="app.posts.post.description"
      />
    }
    value="Lorem Ipsum Lorem Ipsum Lorem Ipsum"
  />
</Card>
`;

exports[`PostDetails renders data unavailable message if post is undefined 1`] = `
<DataUnavailable
  message={
    <FormattedMessage
      id="app.error-no-data"
    />
  }
/>
`;
